internal/impl/cypher/input_cypher.go

```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		return nil, err
+	}
+	uri, err := conf.FieldString(cypherURI)
+	if err != nil {
+		return nil, err
+	}
+	noAuth, err := conf.FieldBool(cypherNoAuth)
+	if err != nil {
+		return nil, err
+	}
+	query, err := conf.FieldString(cypherQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	recordsChan := make(chan *neo4j.Record, 10)
+
+	var tlsEnabled bool
+	var tlsConfig *tls.Config
+	if tlsConfig, tlsEnabled, err = conf.FieldTLSToggled(cypherTLS); err != nil {
+		return nil, err
+	}
+
+	input := &CypherInput{
+		database:    database,
+		uri:         uri,
+		noAuth:      noAuth,
+		query:       query,
+		recordsChan: recordsChan,
+		shutSig:     shutdown.NewSignaller(),
+		logger:      mgr.Logger(),
+	}
+
+	if !noAuth {
+
+		basicAuthMap, err := conf.FieldStringMap(cypherBasicAuth)
+		if err != nil {
+			return nil, err
+		}
+
+		basicAuth := CypherBasicAuth{
+			user:     basicAuthMap[cypherUser],
+			password: basicAuthMap[cypherPassword],
+			realm:    basicAuthMap[cypherRealm],
+		}
+
+		input.basicAuth = basicAuth
+
+		if tlsEnabled {
+			input.tlsConfig = tlsConfig
+		}
+	}
+
+	return input, nil
+}
+
+func (cyp *CypherInput) Connect(ctx context.Context) error {
+
+	var driver neo4j.DriverWithContext
+	var err error
+
+	var connConfigurers []func(*config.Config)
+	if cyp.tlsConfig != nil {
+		connConfigurers = append(connConfigurers, func(config *config.Config) {
+			config.TlsConfig = cyp.tlsConfig
+		})
+	}
+
+	if cyp.noAuth {
+		driver, err = neo4j.NewDriverWithContext(cyp.uri, neo4j.NoAuth())
+	} else {
+		driver, err = neo4j.NewDriverWithContext(cyp.uri, neo4j.BasicAuth(cyp.basicAuth.user, cyp.basicAuth.password, cyp.basicAuth.realm), connConfigurers...)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	cyp.driver = driver
+	cyp.session = driver.NewSession(ctx, neo4j.SessionConfig{AccessMode: neo4j.AccessModeRead})
+
+	go func() {
+		defer close(cyp.recordsChan)
+
+		if _, err := cyp.session.ExecuteRead(ctx,
+			func(tx neo4j.ManagedTransaction) (any, error) {
+				result, err := tx.Run(ctx, cyp.query, nil)
+				if err != nil {
+					cyp.logger.With("err", err).Error("unexpected error while executing cypher query")
+					return nil, err
+				}
+
+				for result.Next(ctx) {
+					select {
+					case <-ctx.Done():
+						return nil, ctx.Err()
+					case <-cyp.shutSig.HardStopChan():
+						return nil, service.ErrEndOfInput
+					case cyp.recordsChan <- result.Record():
+					}
+				}
+				return nil, err
+			}); err != nil {
+			cyp.logger.With("err", err).Error("unexpected error while executing cypher query")
+		}
+	}()
+
+	return nil
+}
+
+func (cyp *CypherInput) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+	if cyp.session == nil && cyp.driver == nil {
+		return nil, nil, service.ErrNotConnected
+	}
+
+	record, ok := <-cyp.recordsChan
+
+	if !ok {
+		return nil, nil, service.ErrEndOfInput
+	}
+
+	msg := service.NewMessage(nil)
+	msg.SetStructuredMut(record.AsMap())
```

Comment:
Think we should be defensive here and do a `select` with the `Read` context:

```suggestion
	msg := service.NewMessage(nil)
	select {
	case record, open := <-cyp.recordsChan:
		if !open {
			return nil, nil, service.ErrEndOfInput
		}
		msg.SetStructuredMut(record.AsMap())
	case <-ctx.Done():
		return nil, nil, service.ErrEndOfInput
	}        	
```

Also, consider locking when we receive the channel from the struct. If there are multiple reads in parallel there's a chance we close the `cyp.recordsChan` and some other process tries to run `cyp.Read` -- which will result in us trying to read in from a closed channel and panic.

See mqtt for reference:

https://github.com/warpstreamlabs/bento/blob/16ebd504192b67b49827ece3ac68aae1d339ce5a/internal/impl/mqtt/input.go#L190-L193

Reply:
I think that this is fine as is, 

reads on a closed buffered channel won't result in a panic - a read on a closed buffered channel will yield the remaining values until the buffer is empty and will then yield a zero value for the channel's type. 

We are using the comma ok idiom to check if the channel is open or closed, and if it is closed we are returning appropriate values. 

we defer the closing of the channel in the go routine started by the Connect function

Reply:
Ahh you're correct on the reading from a closed channel:

> For a channel c, the built-in function close(c) records that no more values will be sent on the channel. After calling close, and after any previously sent values have been received, receive operations will return the zero value for the channel's type without blocking.

Will have to re-look at why these other implementations, like mqtt, lock on receieving that channel. I'll address the `ctx` in another comment

Reply:
> Think we should be defensive here and do a select with the Read context:

Having a `select` here means we can be defensively call `Read` and allow a parent caller to cancel/return early if context is cancelled. Like, the expected/ideal behaviour of these components is that a passed in ctx can be used to determine internal behaviour.

Imagine the scenario where `for result.Next(cypCtx)` is just waiting on new data. A call to `Read` will then block so long as no new records are consumed. What if we're wanting to time out a `Read` request after 5 seconds?

Reply:
added a select in the Read() function - thanks